src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,221 @@
+        /// to place the tags at the end of the editor is selected.
+        /// </summary>
+        private void TextView_ViewportWidthChanged(object sender, EventArgs e)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            if (AdornmentLayer is null)
+            {
+                return;
+            }
+
+            var sourceContainer = TextView.TextBuffer.AsTextContainer();
+            if (sourceContainer is null)
+            {
+                AdornmentLayer.RemoveAllAdornments();
+                return;
+            }
+
+            if (!Workspace.TryGetWorkspace(sourceContainer, out var workspace))
+            {
+                AdornmentLayer.RemoveAllAdornments();
+                return;
+            }
+
+            var document = sourceContainer.GetOpenDocumentInCurrentContext();
+            if (document is null)
+            {
+                AdornmentLayer.RemoveAllAdornments();
+                return;
+            }
+
+            var option = workspace.Options.GetOption(InlineDiagnosticsOptions.Location, document.Project.Language);
+            if (option == InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+            {
+                var normalizedCollectionSpan = new NormalizedSnapshotSpanCollection(TextView.TextViewLines.FormattedSpan);
+                UpdateSpans_CallOnlyOnUIThread(normalizedCollectionSpan, removeOldTags: true);
+            }
+        }
+
+        private void OnClassificationFormatMappingChanged(object sender, EventArgs e)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            if (AdornmentLayer is not null)
+            {
+                foreach (var element in AdornmentLayer.Elements)
+                {
+                    var tag = (InlineDiagnosticsTag)element.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.GetClassificationId(tag.ErrorType));
+                    var format = GetFormat(classificationType);
+                    InlineDiagnosticsTag.UpdateColor(format, element.Adornment);
+                }
+            }
+        }
+
+        private TextFormattingRunProperties GetFormat(IClassificationType classificationType)
+        {
+            return _formatMap.GetTextProperties(classificationType);
+        }
+
+        /// <summary>
+        /// Get the spans located on each line so that it can only display the first one that appears on the line
+        /// </summary>
+        private IDictionary<int, IMappingTagSpan<InlineDiagnosticsTag>> GetSpansOnEachLine(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            var map = new Dictionary<int, IMappingTagSpan<InlineDiagnosticsTag>>();
```

Comment:
consider an approach where the caller creates thsi list (usign PooledDictionary) and add its in here instead.